src/distdl/nn/batch_norm.py

```diff
@@ -0,0 +1,227 @@
+from distdl.nn.sum_reduce import SumReduce
+from distdl.utilities.slicing import compute_start_index
+from distdl.utilities.slicing import compute_stop_index
+from distdl.utilities.slicing import worker_layout
+from distdl.utilities.torch import zero_volume_tensor
+
+
+class DistributedBatchNorm(Module):
+    r"""A distributed batch norm layer.
+
+    Applies Batch Normalization using mini-batch statistics.
+    This layer is a distributed and generalized version of the PyTorch BatchNormNd layers.
+    Currently, parallelism is supported in all dimensions except the feature dimension (dimension 2).
+
+    Parameters
+    ----------
+    P_x :
+        Partition of the input tensor.  Outputs are of the same shape,
+        and therefore re-use the input partition.
+    num_features :
+        Number of features in the input.
+        For example, this should equal C in an input of shape (N, C, L).
+    eps : optional
+        A value added to the denominator for numerical stability.
+        Default is 1e-5.
+    momentum : optional
+        The value used for the running_mean and running_var computation.
+        Can be set to None for cumulative moving average (i.e. simple average).
+        Default is 0.1.
+    affine : optional
+        a boolean value that when set to True, this module has learnable affine parameters.
+        Default is True.
+    track_running_stats : optional
+        a boolean value that when set to True, this module tracks the running mean and variance,
+        and when set to False, this module does not track such statistics and uses batch statistics
+        instead in both training and eval modes if the running mean and variance are None.
+        Default is True.
+    """
+
+    def __init__(self, P_x,
+                 num_features, eps=1e-05, momentum=0.1, affine=True,
+                 track_running_stats=True):
+        super(DistributedBatchNorm, self).__init__()
+        self.num_dimensions = len(P_x.shape)
+        assert self.num_dimensions >= 2
+        self.num_features = num_features
+        self.eps = eps
+        self.momentum = momentum
+        self.affine = affine
+        self.track_running_stats = track_running_stats
+        self.inputs_seen = 0
+
+        # Determine the size of the local trainable parameters (this is a bit of a hack)
+        possible_input_shape = P_x.shape.tolist()
+        possible_input_shape[1] = num_features
+        start_index = compute_start_index(P_x.shape, P_x.index, possible_input_shape)
+        stop_index = compute_stop_index(P_x.shape, P_x.index, possible_input_shape)
+        self.local_num_features = stop_index[1] - start_index[1]
+
+        internal_data_shape = [1] * self.num_dimensions
+        internal_data_shape[1] = self.local_num_features
+        internal_partition_shape = [1] * self.num_dimensions
+        internal_partition_shape[1] = P_x.shape[1]
+
+        # Decide which workers will be used to store sum and affine parameters
```

Comment:
By array, are you referring to the result of `worker_layout()` or `index`?

Reply:
`worker_layout()` returns a large-ish array.  If the (C-ordered) lexicographic ordering is assumed, it would actually be pretty trivial to generate the list of workers from a slice (and the shape of the partition).